src/os/apple/register.rs

```diff
@@ -139,27 +148,85 @@ impl Drop for ServiceRef {
 
 impl RegisteredDnsService {}
 
-// In order to signal the blocked thread, we close its socket to unblock it
+// Signal the background thread via an atomic flag
 impl Drop for RegisteredDnsService {
     fn drop(&mut self) {
-        unsafe {
-            trace!("Closing socket to signal service cleanup...");
-            libc::close(self.socket);
-        }
+        trace!("Signaling thread to exit...");
+        self.shutdown_flag.store(true, Ordering::Release);
     }
 }
-fn run_thread(service: ServiceRef) {
-    std::thread::spawn(move || loop {
-        unsafe {
-            trace!("Processing...");
-            let r = DNSServiceProcessResult(service.raw);
-            if r != kDNSServiceErr_NoError {
-                error!("Error processing: {}, exiting thread", r);
+fn run_thread_with_poll(service: ServiceRef, shutdown_flag: Arc<AtomicBool>) {
+    let socket = unsafe { DNSServiceRefSockFD(service.raw) };
+
+    std::thread::spawn(move || {
+        loop {
+            // Check a shutdown flag
+            if shutdown_flag.load(Ordering::Acquire) {
```

Comment:
Same question with this ordering since it's `Acquire` loading coupled with `Release` storage. It's been a while since I feel like I understood [the docs](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html) but:

> Notice that using this [`Acquire`] ordering for an operation that combines loads and stores leads to a [Relaxed](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Relaxed) store operation!